src/bindings/python/flux/util.py

```diff
@@ -380,3 +382,163 @@ def format(self, obj):
+    """Very simple pstree-like display for the console
+
+    Args:
+        label: Label for this node
+        prefix: Text which comes before connector and label
+        combine_children: combine like children as they are added
+    """
+
+    Connectors = namedtuple("Connectors", ["PIPE", "TEE", "ELBOW", "BLANK"])
+
+    connector_styles = {
+        "ascii": Connectors("|   ", "|-- ", "`-- ", "    "),
+        "box": Connectors("│   ", "├── ", "└── ", "    "),
+        "compact": Connectors("│ ", "├─", "└─", "  "),
+    }
+
+    def __init__(self, label, prefix="", combine_children=False):
+
+        self.label = label
+        self.prefix = prefix
+        if self.prefix:
+            self.prefix += " "
+
+        self.duplicates = {}
+        self.children = []
+        self.count = 1
+        self.combine_children = combine_children
+
+    def append(self, label, prefix=""):
+        """Append a new child, possibly combining duplicates
+
+        Args:
+            label: Label for this node
+            prefix: Optional prefix to display before tree part
+
+        """
+        if not self.combine_children:
+            self.add(label, prefix, combine_children=False)
+            return
+        clabel = f"{prefix}{label}"
+        if clabel in self.duplicates:
+            self.duplicates[clabel].increment()
+        else:
+            tree = self.add(label, prefix, combine_children=True)
+            self.duplicates[clabel] = tree
+
+    def append_tree(self, tree):
+        """Add a child Tree, combining duplicates
+
+        Args:
+            tree: the Tree object to append
+        """
+        if not self.combine_children or tree.children:
+            self.children.append(tree)
+            return
+        clabel = f"{tree.prefix}{tree.label}"
+        if clabel in self.duplicates:
+            self.duplicates[clabel].increment()
+        else:
+            self.children.append(tree)
+            self.duplicates[clabel] = tree
+
+    def add(self, label, prefix="", combine_children=False):
+        """Add a child by prefix and label, does not combine duplicates
+
+        Args:
+            label: Label for this node
+            prefix: Optional prefix to display before tree part
+            combine_children: True if like children should be combined
+
+        Returns:
+            Tree: the new child Tree object
+        """
+        tree = Tree(label, prefix=prefix, combine_children=combine_children)
+        self.children.append(tree)
+        return tree
+
+    def increment(self):
+        self.count = self.count + 1
+
+    def _render(
+        self,
+        pstack=None,
+        connector="",
+        style="box",
+        level=None,
+        truncate=None,
+    ):
+        if pstack is None:
+            pstack = []
+        if level and level < len(pstack):
+            return
+
+        if self.count > 1:
+            label = f"{self.count}*[{self.label}]"
+        else:
+            label = self.label
+
```

Comment:
took me awhile to grok this recursion.  I'm sure it was hard to write :-)

I think the `pstack[:-1]` sort of threw me for a loop b/c we're not using the last piece of the pstack, so it sort of doesn't make sense until you walk yourself threw it a bit.  Like one level of recursion down, the append to the pstack isn't used at all.  

I'm sure some comments would help.

One random idea (possibly not good, but just brainstorming), since the last append of the `pstack` isn't used when you recurse only one level down, instead of appending to `pstack`, how about just pass a variable down.  Like hypothetically:

```
def _render(
        self,
        pstack=None,
        next_in_pstack=None
        connector="",
        style="box",
        level=None,
        truncate=None,
    ):
```

So like if `next_in_pstack != None` then you append to pstack, then set `next_in_pstack` to the next recursive step?

Reply:
Does it help if you think of the stack as the stack of characters required in before the tree connector if you need to recurse another level down? Your suggestion to pass the next character required as a separate argument seems good, but I don't see much difference between the current approach and that (except adding more parameters to the function). Perhaps some comments will help.

Reply:
Yeah, some comments would help.  The way I eventually thought about it is that you are passing a connector and an additional prefix into the next level of recursion.  It just so happens that the additional prefix is never used in the next recursive level, but is only used 2 recursions later.

Another brainstorming idea (again, possibly not good), you could do something like:

```
if pstack[-1] == connectors.TEE
   pstack[-1] = connectors.PIPE
pstack.append(connects.TEE)
```

kinda thing in the recursive step so that the whole `pstack` is just output everytime you recurse.  I'm not sure if it's a net win logic wise or not.  Perhaps simplifying the output code but complicating the recursive step.

Reply:
I don't know, it feels like we're getting into a matter of opinion and style here and I'd prefer to keep the code the way I wrote it since I understand that approach a little better. Let me push my fixup commit with additional comments and see what you think.

Reply:
acknowledged, its a style thing at this point.  The comments you added definitely help.